### PR TITLE
fix(cli): output more detailed information for steps when using JUnit reporter

### LINKED ISSUE #22795: JUnit reporter: Output XML doesn't contain any details regarding failures on test "step"s
## Issue

We can write nested test cases using `t.step` (where `t` is [`TestContext`]).
When we run `deno test --reporter=junit` for a test file containing nested test cases that fail, the output XML will not contain useful information such as the place where it failed or how it failed (e.g. assertion messages).

[`TestContext`]: https://deno.land/api@v1.41.2?s=Deno.TestContext

### Deno

Say we have the following test file:

```ts
import { assertEquals } from "https://deno.land/std@0.217.0/assert/mod.ts";

Deno.test({
  name: "parent",
  fn: async (t) => {
    await t.step("step1", () => {
      assertEquals(42, 5); // should fail
    });
    await t.step("step2", () => {
      assertEquals(42, 42); // should succeed
    });
  },
});
```

Running `deno test --reporter=junit` gets us:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="deno test" tests="1" failures="1" errors="0" time="0.003">
    <testsuite name="file:///private/tmp/tmpmar8/main_test.ts" tests="1" disabled="0" errors="0" failures="1">
        <testcase name="parent" time="0.002" filename="/private/tmp/tmpmar8/main_test.ts" line="3" col="6">
            <properties>
                <property name="step[failure]" value="step1"/>
                <property name="step[passed]" value="step2"/>
            </properties>
            <failure message="1 test step failed."/>
        </testcase>
    </testsuite>
</testsuites>
```

We can tell the fact that `step1` failed and `step2` succeeded, but no other information is available. In particular we would like to get assertion error messages like `Actual 42, Got 5` or something.

### Other test runners

Take a look at vitest for comparison. 

Here's a input to provide to vitest. Note that the structure is the same as we did above for `deno test`.

```ts
import { describe, it } from "vitest";

describe("parent", () => {
  it("step1", ({ expect }) => {
    expect(42).toBe(5); // should fail
  });
  it("step2", ({ expect }) => {
    expect(42).toBe(42); // should succeed
  });
});
```

Running vitest with its JUnit reporter enabled generates the following output:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<testsuites name="vitest tests" tests="2" failures="1" errors="0" time="0.121">
    <testsuite name="vitest.test.ts" timestamp="2024-03-08T12:39:17.635Z" hostname="Yusukes-MacBook-Air.local" tests="2" failures="1" errors="0" skipped="0" time="0.004">
        <testcase classname="vitest.test.ts" name="parent &gt; step1" time="0.003">
            <failure message="expected 42 to be 5 // Object.is equality" type="AssertionError">
AssertionError: expected 42 to be 5 // Object.is equality
 ❯ vitest.test.ts:5:16
            </failure>
        </testcase>
        <testcase classname="vitest.test.ts" name="parent &gt; step2" time="0.001">
        </testcase>
    </testsuite>
</testsuites>
```

This shows that the failed test step gives an assertion error message and where it failed, which is pretty nice.

My proposal here is that we improve Deno's JUnit reporter so it can output as much information as vitest.

## Version

Version: Deno 1.41.2

### PR DESCRIPTION
**Pull request by magurotuna**
This patch gets JUnit reporter to output more detailed information for test steps (subtests).

## Issue with previous implementation

In the previous implementation, the test hierarchy was represented using several XML tags like the following:

- `<testsuites>` corresponds to the entire test (one execution of `deno test` has exactly one `<testsuites>` tag)
- `<testsuite>` corresponds to one file, such as `main_test.ts`
- `<testcase>` corresponds to one `Deno.test(...)`
- `<property>` corresponds to one `t.step(...)`

This structure describes the test layers but one problem is that `<property>` tag is used for any use cases so some tools that can ingest a JUnit XML file might not be able to interpret `<property>` as subtests. 

## How other tools address it

Some of the testing frameworks in the ecosystem address this issue by fitting subtests into the `<testcase>` layer. For instance, take a look at the following Go test file:

```go
package main_test

import "testing"

func TestMain(t *testing.T) {
  t.Run("child 1", func(t *testing.T) {
    // OK
  })

  t.Run("child 2", func(t *testing.T) {
    // Error
    t.Fatal("error")
  })
}
```

Running [gotestsum], we can get the output like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="3" failures="2" errors="0" time="1.013694">
	<testsuite tests="3" failures="2" time="0.510000" name="example/gosumtest" timestamp="2024-03-11T12:26:39+09:00">
		<properties>
			<property name="go.version" value="go1.22.1 darwin/arm64"></property>
		</properties>
		<testcase classname="example/gosumtest" name="TestMain/child_2" time="0.000000">
			<failure message="Failed" type="">=== RUN   TestMain/child_2&#xA;    main_test.go:12: error&#xA;--- FAIL: TestMain/child_2 (0.00s)&#xA;</failure>
		</testcase>
		<testcase classname="example/gosumtest" name="TestMain" time="0.000000">
			<failure message="Failed" type="">=== RUN   TestMain&#xA;--- FAIL: TestMain (0.00s)&#xA;</failure>
		</testcase>
		<testcase classname="example/gosumtest" name="TestMain/child_1" time="0.000000"></testcase>
	</testsuite>
</testsuites>
``` 

This output shows that nested test cases are squashed into the `<testcase>` layer by treating them as the same layer as their parent, `TestMain`. We can still distinguish nested ones by their `name` attributes that look like `TestMain/<subtest_name>`.

As described in #22795, [vitest] solves the issue in the same way as [gotestsum].

One downside of this would be that one test failure that happens in a nested test case will end up being counted multiple times, because not only the subtest but also its wrapping container(s) are considered to be failures. In fact, in the [gotestsum] output above, `TestMain/child_2` failed (which is totally expected) while its parent, `TestMain`, was also counted as failure. As https://github.com/denoland/deno/pull/20273#discussion_r1307558757 pointed out, there is a test runner that offers flexibility to prevent this, but I personally don't think the "duplicate failure count" issue is a big deal.

## How to fix the issue in this patch

This patch fixes the issue with the same approach as [gotestsum] and [vitest].
More specifically, nested test cases are put into the `<testcase>` level and their names are now represented as squashed test names concatenated by `>` (e.g. `parent 2 > child 1 > grandchild 1`). This change also allows us to put a detailed error message as `<failure>` tag within the `<testcase>` tag, which should be handled nicely by third-party tools supporting JUnit XML.

## Extra fix

Also, file paths embedded into XML outputs are changed from absolute path to relative path, which is helpful when running the test suites in several different environments like CI.

Resolves #22795

[gotestsum]: https://github.com/gotestyourself/gotestsum
[vitest]: https://vitest.dev/

### CODE REVIEW

File: cli/tools/test/mod.rs
Comment (magurotuna) [inline cli/tools/test/mod.rs:271]: For reviewers: this is just a refactor; `ToString` trait is automatically implemented for all types that implement `std::fmt::Display`, so basically it's recommended to implement `std::fmt::Display`

**Comment by bartlomieju** [comment]
> One downside of this would be that one test failure that happens in a nested test case will end up being counted multiple times, because not only the subtest but also its wrapping container(s) are considered to be failures. In fact, in the [gotestsum](https://github.com/gotestyourself/gotestsum) output above, TestMain/child_2 failed (which is totally expected) while its parent, TestMain, was also counted as failure. As https://github.com/denoland/deno/pull/20273#discussion_r1307558757 pointed out, there is a test runner that offers flexibility to prevent this, but I personally don't think the "duplicate failure count" issue is a big deal.

I agree, this is a reasonable trade-off.

File: cli/tools/test/reporters/junit.rs
Comment (bartlomieju) [inline cli/tools/test/reporters/junit.rs:243]: Maybe pre-allocate to something like 256 to avoid excessive reallocations?

**Review by bartlomieju — COMMENTED** [review]
Great improvement. I have a couple of requests, but overall LGTM.

Pinging @littledivy for a second pair of eyes and to ensure `npm_smoke_tests` repo won't suddenly break because of this change.

File: cli/tools/test/reporters/junit.rs
Comment (magurotuna) [reply cli/tools/test/reporters/junit.rs:23]: I brought this from the existing code of PrettyReporter:

https://github.com/denoland/deno/blob/b6e44f91ad55f9737d65a4832d10cfa608f27c41/cli/tools/test/reporters/pretty.rs#L40

Should we also fix these ones as well?

File: cli/tools/test/reporters/junit.rs
Comment (magurotuna) [reply cli/tools/test/reporters/junit.rs]: Good point! fixed in [14d89d2](https://github.com/denoland/deno/pull/22797/commits/14d89d281015b73b99c7cb0304b4b1b04e8ee112)

File: cli/tools/test/reporters/junit.rs
Comment (bartlomieju) [reply cli/tools/test/reporters/junit.rs:23]: Yes, I think it would be better to fix both then

Comment (magurotuna) [reply cli/tools/test/reporters/junit.rs]: OK, it's done [752db03](https://github.com/denoland/deno/pull/22797/commits/752db030714e90b016d07c3a925eed6de6a736b8)

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**
